js/ai/src/reranker.ts

```diff
@@ -0,0 +1,244 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Action, defineAction } from '@genkit-ai/core';
+import { lookupAction } from '@genkit-ai/core/registry';
+import * as z from 'zod';
+import { Part, PartSchema } from './document.js';
+import { Document, DocumentData, DocumentDataSchema } from './retriever.js';
+
+type RerankerFn<RerankerOptions extends z.ZodTypeAny> = (
+  query: Document,
+  documents: Document[],
+  queryOpts: z.infer<RerankerOptions>
+) => Promise<RerankerResponse>;
+
+export const RankedDocumentDataSchema = z.object({
+  content: z.array(PartSchema),
+  metadata: z
+    .object({
+      score: z.number(), // Enforces that 'score' must be a number
+    })
+    .passthrough(), // Allows other properties in 'metadata' with any type
+});
+
+export type RankedDocumentData = z.infer<typeof RankedDocumentDataSchema>;
+
+export class RankedDocument extends Document implements RankedDocumentData {
+  content: Part[];
+  metadata: { score: number } & Record<string, any>;
+
+  constructor(data: RankedDocumentData) {
+    super(data);
+    this.content = data.content;
+    this.metadata = data.metadata;
+  }
+
+  static fromText(
+    text: string,
+    metadata: { score: number } & Record<string, any>
+  ) {
+    // TODO: do we need validation here? I'm thinking in javascript we won't have type errors
+    if (metadata.score === undefined) {
+      throw new Error('Score is required');
+    }
+    return new RankedDocument({
+      content: [{ text }],
+      metadata,
+    });
+  }
+
+  /**
+   * Concatenates all `text` parts present in the document with no delimiter.
+   * @returns A string of all concatenated text parts.
+   */
+  text(): string {
```

Comment:
Document already implements this - remove it

Reply:
In fact i've reimplemented a bunch of things here that i dont have to. I will remove.